src/xwayland.c

```diff
@@ -580,6 +596,62 @@ handle_set_strut_partial(struct wl_listener *listener, void *data)
 	}
 }
 
+static void
+update_icon(struct xwayland_view *xwayland_view)
+{
+	if (!xwayland_view->xwayland_surface) {
+		return;
+	}
+
+	xcb_window_t window_id = xwayland_view->xwayland_surface->window_id;
+
+	xcb_connection_t *xcb_conn = wlr_xwayland_get_xwm_connection(
+		xwayland_view->base.server->xwayland);
+	xcb_get_property_cookie_t cookie = xcb_get_property(xcb_conn, 0,
+		window_id, atoms[ATOM_NET_WM_ICON], XCB_ATOM_CARDINAL, 0, 0x10000);
+	xcb_get_property_reply_t *reply = xcb_get_property_reply(xcb_conn, cookie, NULL);
+	if (!reply) {
+		return;
+	}
+	xcb_ewmh_get_wm_icon_reply_t icon;
+	if (!xcb_ewmh_get_wm_icon_from_reply(&icon, reply)) {
+		wlr_log(WLR_INFO, "Invalid x11 icon");
+		view_set_icon(&xwayland_view->base, NULL, NULL);
+		goto out;
+	}
+
+	xcb_ewmh_wm_icon_iterator_t iter = xcb_ewmh_get_wm_icon_iterator(&icon);
+	struct wl_array buffers;
+	wl_array_init(&buffers);
+	for (; iter.rem; xcb_ewmh_get_wm_icon_next(&iter)) {
+		size_t stride = iter.height * 4;
+		uint32_t *buf = calloc(1, iter.width * stride);
+
+		/* Pre-multiply alpha */
+		for (uint32_t y = 0; y < iter.height; y++) {
+			for (uint32_t x = 0; x < iter.width; x++) {
+				uint32_t i = x + y * iter.width;
+				uint8_t *src_pixel = (uint8_t *)&iter.data[i];
+				uint8_t *dst_pixel = (uint8_t *)&buf[i];
+				dst_pixel[0] = src_pixel[0] * src_pixel[3] / 255;
+				dst_pixel[1] = src_pixel[1] * src_pixel[3] / 255;
+				dst_pixel[2] = src_pixel[2] * src_pixel[3] / 255;
+				dst_pixel[3] = src_pixel[3];
+			}
+		}
+
+		struct lab_data_buffer *buffer = buffer_create_from_data(
+			buf, iter.width, iter.height, stride);
+		array_add(&buffers, buffer);
+	}
+
```

Comment:
Suggest adding a comment like `/* view takes ownership of the buffers */ to make clear that the `calloc()` call above doesn't leak although it might appear like it.